compiler/packages/babel-plugin-react-compiler/src/Transform/TransformFire.ts

```diff
@@ -0,0 +1,625 @@
+              );
+              newInstrs.push(
+                loadUseFireInstr,
+                loadFireCalleeInstr,
+                callUseFireInstr,
+                storeUseFireInstr,
+              );
+
+              // We insert all of these instructions before the useEffect is loaded
+              const loadUseEffectInstrId = context.getLoadGlobalInstrId(
+                value.callee.identifier.id,
+              );
+              if (loadUseEffectInstrId == null) {
+                context.pushError({
+                  loc: value.loc,
+                  description: null,
+                  severity: ErrorSeverity.Invariant,
+                  reason: '[InsertFire] No LoadGlobal found for useEffect call',
+                  suggestions: null,
+                });
+                continue;
+              }
+              rewriteInstrs.set(loadUseEffectInstrId, newInstrs);
+            }
+          }
+        }
+      } else if (
+        value.kind === 'CallExpression' &&
+        value.callee.identifier.type.kind === 'Function' &&
+        value.callee.identifier.type.shapeId === BuiltInFireId &&
+        context.inUseEffectLambda()
+      ) {
+        /*
+         * We found a fire(callExpr()) call. We remove the `fire()` call and replace the callExpr()
+         * with a freshly generated fire function binding. We'll insert the useFire call before the
+         * useEffect call, which happens in the CallExpression (useEffect) case above.
+         */
+
+        /*
+         * We only allow fire to be called with a CallExpression: `fire(f())`
+         * TODO: add support for method calls: `fire(this.method())`
+         */
+        if (value.args.length === 1 && value.args[0].kind === 'Identifier') {
+          const callExpr = context.getCallExpression(
+            value.args[0].identifier.id,
+          );
+
+          if (callExpr != null) {
+            const calleeId = callExpr.callee.identifier.id;
+            const loadLocal = context.getLoadLocalInstr(calleeId);
+            if (loadLocal == null) {
+              context.pushError({
+                loc: value.loc,
+                description: null,
+                severity: ErrorSeverity.Invariant,
+                reason:
+                  '[InsertFire] No loadLocal found for fire call argument',
+                suggestions: null,
+              });
+              continue;
+            }
+
+            const fireFunctionBinding =
+              context.getOrGenerateFireFunctionBinding(
+                {...loadLocal.place},
+                fn.env,
+              );
+
+            loadLocal.place = {...fireFunctionBinding};
+
+            // Delete the fire call expression
+            deleteInstrs.add(instr.id);
+          } else {
+            context.pushError({
+              loc: value.loc,
+              description:
+                '`fire()` can only receive a `callExpression()`. Method calls and other expressions are not allowed',
```

Comment:
nit: `callExpression()` refers to an ast name, let's use a more user-friendly name like "can only receive a function call such as `fire(fn(a, b))`"